lib/Optimizer/Dialect/CC/CCOps.cpp

```diff
@@ -187,10 +187,73 @@ void cudaq::cc::AllocaOp::getCanonicalizationPatterns(
 // CastOp
 //===----------------------------------------------------------------------===//
 
-OpFoldResult cudaq::cc::CastOp::fold(FoldAdaptor) {
+OpFoldResult cudaq::cc::CastOp::fold(FoldAdaptor adaptor) {
   // If cast is a nop, just forward the argument to the uses.
   if (getType() == getValue().getType())
     return getValue();
+  if (auto optConst = adaptor.getValue()) {
+    // Replace a constant + cast with a new constant of an updated type.
+    auto ty = getType();
+    OpBuilder builder(*this);
+    auto fltTy = builder.getF32Type();
+    auto dblTy = builder.getF64Type();
+    auto loc = getLoc();
+    if (auto attr = dyn_cast<IntegerAttr>(optConst)) {
+      auto val = attr.getInt();
+      if (isa<IntegerType>(ty)) {
+        auto width = ty.getIntOrFloatBitWidth();
+        return builder.create<arith::ConstantIntOp>(loc, val, width)
+            .getResult();
+      } else if (ty == fltTy) {
+        if (getZint()) {
+          APFloat fval(static_cast<float>(static_cast<std::uint64_t>(val)));
+          return builder.create<arith::ConstantFloatOp>(loc, fval, fltTy)
+              .getResult();
+        }
+        if (getSint()) {
+          APFloat fval(static_cast<float>(val));
+          return builder.create<arith::ConstantFloatOp>(loc, fval, fltTy)
+              .getResult();
+        }
+      } else if (ty == dblTy) {
+        if (getZint()) {
+          APFloat fval(static_cast<double>(static_cast<std::uint64_t>(val)));
+          return builder.create<arith::ConstantFloatOp>(loc, fval, dblTy)
+              .getResult();
+        }
+        if (getSint()) {
+          APFloat fval(static_cast<double>(val));
+          return builder.create<arith::ConstantFloatOp>(loc, fval, dblTy)
+              .getResult();
+        }
+      }
+    }
+    if (auto attr = dyn_cast<FloatAttr>(optConst)) {
+      auto val = attr.getValue();
+      if (isa<IntegerType>(ty)) {
+        auto width = ty.getIntOrFloatBitWidth();
+        if (getZint()) {
+          std::uint64_t v = val.convertToDouble();
+          return builder.create<arith::ConstantIntOp>(loc, v, width)
+              .getResult();
+        }
+        if (getSint()) {
+          std::int64_t v = val.convertToDouble();
+          return builder.create<arith::ConstantIntOp>(loc, v, width)
+              .getResult();
```

Comment:
Would there be any harm in doing this?
```suggestion
          if (auto d = val.convertToDouble(); d >= 0.0) {
            std::int64_t v = d;
            return builder.create<arith::ConstantIntOp>(loc, v, width)
                .getResult();
          }
```

Reply:
The reason I ask is that if the different platforms behave differently due to the underlying hardware implementations, the compilers may be stuck with that, but with constant folding, it seems like there is no benefit to being stuck with a bad cast if it can simply be detected and avoided.

Reply:
Casting floating-point values to integers is rife with undefined behavior. It isn't the job of `nvq++` to protect a programmer from C++'s undefined behavior as far as I understand it. Adding error checks for undefined behavior doesn't seem like a good use of our limited resources, in my opinion.

Again, this isn't a bug. Or, if you'd rather, `nvq++` is faithfully "bug for bug compatible" in this regard.

The only other option is to drop constant folding floating-point to integral casts completely here, which seems like a mistake to some degree. On the other hand, it will mean LLVM will do the exact same constant folding later should this code go through LLVM.